experimental/ActionPolyRing/src/Types.jl

```diff
@@ -0,0 +1,203 @@
+    jet_to_upoly_idx = Dict{Tuple{Int, Vector{Int}}, Int}()
+    
+    return new{T}(upoly_ring, elementary_symbols, ndiffs, false, jet_to_var, var_to_jet, jet_to_upoly_idx)
+  end
+
+end
+
+mutable struct DifferentialPolyRingElem{T} <: ActionPolyRingElem{T}
+  p::AbstractAlgebra.Generic.UniversalPolyRingElem{T}
+  parent::DifferentialPolyRing{T}
+  initial::Any #Always of type DifferentialPolyRingElem{T}
+  permutation::Vector{Int}
+
+  DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}) where {T} = new{T}(zero(dpr.upoly_ring), dpr)
+
+  function DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}, upre::AbstractAlgebra.Generic.UniversalPolyRingElem{T}) where {T}
+    @req dpr.upoly_ring === parent(upre) "The parent does not match"
+    new{T}(upre, dpr)
+  end
+
+  function DifferentialPolyRingElem{T}(dpr::DifferentialPolyRing{T}, mpre::MPolyRingElem{T}) where {T}
+    @req dpr.upoly_ring.mpoly_ring === parent(mpre) "The parent does not match"
+    new{T}(mpre, dpr)
+  end
+
+end
+
+struct DifferentialPolyCoeffs{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyExponentVectors{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyTerms{T <: RingElem}
+   poly::T
+end
+
+struct DifferentialPolyMonomials{T <: RingElem}
+   poly::T
+end
+
+###############################################################################
+#
+#  Rankings 
+#
+###############################################################################
+
+# For a ranking of jet variables, e.g. u1[123], u3[041], ... one needs:
+# - An ordering to compare the elementary symbols u_1, ..., u_n
+# - An ordering of the multiindices (this is like a monomial_ordering)
+# - A decision on which of the two has priority, e.g. position-over-term (pot)
+#   or term-over-position (top). Here, the position of the jet variable corresponding
+#   to (i, [a_1, ..., a_n]) is the first coordinate, i.e. i. More generally, this decision means
+#   choosing an ordered partition of the set {1, ...,m}, pot and top being the choices of the two trivial
+#   partitions
+
+abstract type Ranking end
+
+### Difference ###
+mutable struct DifferenceRanking{T} <: Ranking where {T}
+  ring::DifferencePolyRing{T}
+  partition::Vector{Vector{Int}}
+  index_ordering_matrix::ZZMatrix
+  riquier_matrix::ZZMatrix
+
+  function DifferenceRanking{T}(dpr::DifferencePolyRing{T}, partition::Vector{Vector{Int}}, index_ordering_matrix::ZZMatrix) where {T}  
+    return new{T}(dpr, partition, index_ordering_matrix)
+  end
+  
+end
+
+### Differential ###
+mutable struct DifferentialRanking{T} <: Ranking where {T}
+  ring::DifferentialPolyRing{T}
+  partition::Vector{Vector{Int}}
+  index_ordering_matrix::ZZMatrix
+  riquier_matrix::ZZMatrix
+
+  function DifferentialRanking{T}(dpr::DifferentialPolyRing{T}, partition::Vector{Vector{Int}}, index_ordering_matrix::ZZMatrix) where {T}  
+    return new{T}(dpr, partition, index_ordering_matrix)
+  end
+  
+end
```

Comment:
These two structs look completely the same, so it would probably be possible to merge them, and have the ring as a type param instead. So something like (untested):
```suggestion
mutable struct ActionPolyRingRanking{PolyT} <: Ranking where {T, PolyT <: ActionPolyRing{T}}
  ring::PolyT
  partition::Vector{Vector{Int}}
  index_ordering_matrix::ZZMatrix
  riquier_matrix::ZZMatrix

  function ActionPolyRingRanking{PolyT}(dpr::PolyT, partition::Vector{Vector{Int}}, index_ordering_matrix::ZZMatrix) where {T, PolyT <: ActionPolyRing{T}}
    return new{PolyT}(dpr, partition, index_ordering_matrix)
  end
  
end
```

Reply:
This seems to work. At least my tests say so.